rust/otel-arrow-rust/src/otlp/attributes/decoder.rs

```diff
@@ -62,3 +76,532 @@ where
+                        .append_value(parent_id_arr.value_at_or_default(batch_idx));
+                }
+            }
+
+            curr_range_start = idx + 1;
+        }
+    }
+    let materialized_parent_ids = Arc::new(materialized_parent_ids.finish());
+
+    // create new record batch but with parent column replaced
+    let parent_id_idx = record_batch
+        .schema()
+        .index_of(consts::PARENT_ID)
+        .expect("parent_id should be in the schema");
+    let columns = record_batch
+        .columns()
+        .iter()
+        .enumerate()
+        .map(move |(i, col)| {
+            if i == parent_id_idx {
+                materialized_parent_ids.clone()
+            } else {
+                col.clone()
+            }
+        })
+        .collect::<Vec<ArrayRef>>();
+
+    // safety: RecordBatch::try_new will only return error if our schema doesn't
+    // match the passed arrays, or if the arrays are different lengths. Both of
+    // these criteria should be met at this point
+    Ok(RecordBatch::try_new(record_batch.schema(), columns)
+        .expect("should be able to create record batch with parent_id replaced"))
+}
+
+// Creates a boolean array where an element having value true means that the
+// element at index i of the passed array equals the element at index i + 1.
+// Nulls are always treated as not equal
+//
+// For example,
+//    arr = [1, 1, 1, 2, 2,    null, null, 1, 1]
+// result = [T, T, F, T, null, null, null, T]
+//
+fn create_next_element_equality_array(arr: &ArrayRef) -> Result<BooleanArray> {
+    // if the array is a dictionary, we compare the dicitonary keys
+    if let DataType::Dictionary(dict_key_type, _) = arr.data_type() {
+        match **dict_key_type {
+            DataType::UInt8 => {
+                let dict = arr
+                    .as_any()
+                    .downcast_ref::<DictionaryArray<UInt8Type>>()
+                    .expect("array can be downcast to DictionaryArray<UInt8Type>");
+
+                Ok(create_next_eq_array_for_array(dict.keys()))
+            }
+            DataType::UInt16 => {
+                let dict = arr
+                    .as_any()
+                    .downcast_ref::<DictionaryArray<UInt16Type>>()
+                    .expect("array can be downcast to DictionaryArray<UInt16Type>");
+                Ok(create_next_eq_array_for_array(dict.keys()))
+            }
+            _ => error::UnsupportedDictionaryKeyTypeSnafu {
+                expect_oneof: vec![DataType::UInt8, DataType::UInt16],
+                actual: (**dict_key_type).clone(),
+            }
+            .fail(),
+        }
+    } else {
+        // array is not a dictionary, so just compare the values directly
+        Ok(create_next_eq_array_for_array(arr))
+    }
+}
+
+fn create_next_eq_array_for_array<T: Array>(arr: T) -> BooleanArray {
+    // use the arrow compute kernel to compare one
+    let lhs = arr.slice(0, arr.len() - 1);
+    let rhs = arr.slice(1, arr.len() - 1);
```

Comment:
Should this be `arr.slice(1, arr.len())`?

Reply:
In this case, the 2nd argument is the length of the slice, not the end index:
https://arrow.apache.org/rust/arrow/array/trait.Array.html#tymethod.slice